beacon-chain/db/block_header.go

```diff
@@ -0,0 +1,106 @@
+package db
+
+import (
+	"bytes"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+
+	"github.com/boltdb/bolt"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+)
+
+func createBlockHeader(enc []byte) (*ethpb.BeaconBlockHeader, error) {
+	protoBlockHead := &ethpb.BeaconBlockHeader{}
+	err := proto.Unmarshal(enc, protoBlockHead)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal encoding")
+	}
+	return protoBlockHead, nil
+}
+
+// BlockHeader accepts a block root and returns the corresponding block.
+// Returns nil if the block does not exist.
+func (db *BeaconDB) BlockHeader(epoch uint64, validatorID uint64) ([]*ethpb.BeaconBlockHeader, error) {
+	var bha []*ethpb.BeaconBlockHeader
+	err := db.view(func(tx *bolt.Tx) error {
+		c := tx.Bucket(historicBlockHeadersBucket).Cursor()
+		prefix := encodeEpochValidatorID(epoch, validatorID)
+		for k, v := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, v = c.Next() {
+			bh, err := createBlockHeader(v)
+			if err != nil {
+				return err
+			}
+			bha = append(bha, bh)
+		}
+		return nil
+	})
+	return bha, err
+}
+
+// HasBlockHeader accepts an epoch and validator id and returns true if the block header exists.
+func (db *BeaconDB) HasBlockHeader(epoch uint64, validatorID uint64) bool {
+	prefix := encodeEpochValidatorID(epoch, validatorID)
+	var hasBlockHeader bool
+	// #nosec G104
+	_ = db.view(func(tx *bolt.Tx) error {
+		c := tx.Bucket(historicBlockHeadersBucket).Cursor()
+		for k, _ := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, _ = c.Next() {
+			hasBlockHeader = true
+			return nil
+		}
+		hasBlockHeader = false
+		return nil
+	})
+
+	return hasBlockHeader
+}
+
+// SaveBlockHeader accepts a block header and writes it to disk.
+func (db *BeaconDB) SaveBlockHeader(epoch uint64, validatorID uint64, blockHeader *ethpb.BeaconBlockHeader) error {
+	key := encodeEpochValidatorIDSig(epoch, validatorID, blockHeader.Signature)
+	enc, err := proto.Marshal(blockHeader)
+	if err != nil {
+		return errors.Wrap(err, "failed to encode block")
+	}
+
+	return db.update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(historicBlockHeadersBucket)
```

Comment:
why do we need a new bucket? `historicBlockHeadersBucket`  why can't we just use epoch and validator id as references? this means we are storing the blocks twice with unnecessary overhead

Reply:
- this bucket doesn't store blocks, it stores block headers.
- according to the design we are going to have a feature flag for running a slasher service. the default config otherwise will be to store 10-20 epochs of block headers